java/src/main/java/ai/rapids/cudf/Aggregation.java

```diff
@@ -385,6 +386,36 @@ public boolean equals(Object other) {
         }
     }
 
+    static final class HostUDFAggregation extends Aggregation {
+        private final long udfNativeHandle;
+
+        private HostUDFAggregation(long udfNativeHandle) {
+            super(Kind.HOST_UDF);
+            this.udfNativeHandle = udfNativeHandle;
+        }
+
+        @Override
+        long createNativeInstance() {
+            return Aggregation.createHostUDFAgg(udfNativeHandle);
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * kind.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) {
+                return true;
+            } else if (other instanceof HostUDFAggregation) {
+                HostUDFAggregation o = (HostUDFAggregation) other;
+                return Aggregation.areHostUDFsEqual(udfNativeHandle, o.udfNativeHandle);
```

Comment:
Call into native function,  is there a way to avoid calling into native `areHostUDFsEqual` function? other types of `Aggregation` did not call into native function. 
Maybe:
Add sub type into HostUDFAggregation, e.g.: for HLLPP, sub type is `HLLPP`. Like the `enum Kind` in `Aggregation.java`.

```java
enum SubHostUDFType {
    HLLPP(1),
    HLLPP_MERGE(2),
    Example(3);

    final int nativeId;

    SubHostUDFType(int nativeId) {
        this.nativeId = nativeId;
    }
}

class HostUDFAggregation {
    long udfNativeHandle;
    int subUDFType;
    Map<String, String> parameters;

    // use SubHostUDFFactory to create
    private HostUDFAggregation(
            long udfNativeHandle,
            int subUDFType,
            Map<String, String> parameters) {
        this.udfNativeHandle = udfNativeHandle;
        this.subUDFType = subUDFType;
        this.parameters = parameters;
    }

    @Override
    public int hashCode() {
        return Objects.hash(Kind.HOST_UDF, subUDFType, parameters);
    }

    public boolean equals(Object o) {
        return o instanceof HostUDFAggregation &&
                Objects.equals(subUDFType, ((HostUDFAggregation) o).subUDFType) &&
                Objects.equals(parameters, ((HostUDFAggregation) o).parameters);
    }

    static abstract class SubHostUDFFactory {
        // call to native
        abstract long createNativeInstance();

        abstract Map<String, String> parameters();

        abstract SubHostUDFType subType();

        public HostUDFAggregation createSubUDF() {
            return new HostUDFAggregation(
                    createNativeInstance(),
                    subType().nativeId,
                    parameters());
        }
    }
}

class GroupByAggregation {
    public GroupByAggregation(HostUDFAggregation udf){}
    GroupByAggregation hostUDF(HostUDFAggregation.SubHostUDFFactory subFactory) {
        return new GroupByAggregation(subFactory.createSubUDF());
    }
}

class HLLPPFactory extends HostUDFAggregation.SubHostUDFFactory {
    private Map<String, String> parameters;

    HLLPPFactory(Map<String, String> parameters) {
        // copy to avoid external change
        parameters = new HashMap<>(parameters);
    }

    // call to native
    @Override
    public long createNativeInstance() {
        int precision = Integer.parseInt(parameters.get("precision"));
        return createNativeInstanceImpl(precision);
    }

    @Override
    Map<String, String> parameters() {
        return parameters;
    }

    @Override
    SubHostUDFType subType() {
        return SubHostUDFType.HLLPP;
    }

    native long createNativeInstanceImpl(int precision);
}
```

Reply:
Yes, I also don't like calling the native function to check for equality. However, here cudf (Java) cannot know how the native UDF is implemented. As such, the two native instances can be checked for equality using many ways that we cannot control, such as:
 * Using just their types (this way can be implemented as subtype in Java, as suggested above), 
 * Using their internal parameter values (two instances of the same subtype are different if they hold different parameter values), 
 * Using a runtime evaluation on some boolean flags (?),
 * etc.

As such, we cannot anticipate and support all these possible ways in cudf Java. We can only rely on the native implementation. Such native calls should not be very expensive to have a significant impact on the performance but I may be wrong. @revans2 what do you think about this?

Reply:
Oh wait, I have an idea. We can use a Java proxy class for equality comparison.

Reply:
Done via https://github.com/rapidsai/cudf/pull/17592/commits/8bc63437cfe0725fe4ea56c232d74764cb6d2dd6